mhcflurry/dataset.py

```diff
@@ -0,0 +1,554 @@
+
+    Design considerations:
+        - want to allow multiple measurements for each pMHC pair (which can
+          be dynamically combined)
+        - optional sample weights associated with each pMHC measurement
+    """
+    def __init__(self, df):
+        """
+        Constructs a DataSet from a pandas DataFrame with the following
+        columns:
+            - allele
+            - peptide
+            - affinity
+
+        Also, there is an optional column:
+            - sample_weight
+
+        If `sample_weight` is missing then it is filled with a default value
+        of 1.0
+
+        Parameters
+        ----------
+        df : pandas.DataFrame
+        """
+        columns = set(df.columns)
+
+        for expected_column_name in {"allele", "peptide", "affinity"}:
+            if expected_column_name not in columns:
+                raise ValueError("Missing column '%s' from DataFrame")
+        # make allele and peptide columns the index, and copy it
+        # so we can add a column without any observable side-effect in
+        # the calling code
+        df = df.set_index(["allele", "peptide"], drop=False)
+
+        if "sample_weight" not in columns:
+            df["sample_weight"] = np.ones(len(df), dtype=float)
+
+        self._df = df
+        self._alleles = np.asarray(df["allele"])
+        self._peptides = np.asarray(df["peptide"])
+        self._affinities = np.asarray(df["affinity"])
+        self._sample_weights = np.asarray(df["sample_weight"])
+
+    def to_dataframe(self):
+        """
+        Returns DataFrame representation of data contained in Dataset
+        """
+        return self._df
+
+    @property
+    def peptides(self):
+        """
+        Array of peptides from pMHC measurements.
+        """
+        return self._peptides
+
+    @property
+    def alleles(self):
+        """
+        Array of MHC allele names from pMHC measurements.
+        """
+        return self._alleles
+
+    @property
+    def affinities(self):
+        """
+        Array of affinities from pMHC measurements.
+        """
+        return self._affinities
+
+    @property
+    def sample_weights(self):
+        """
+        Array of sample weights for each pMHC measurement.
+        """
+        return self._sample_weights
+
+    def __len__(self):
+        return len(self.to_dataframe())
+
+    def __str__(self):
+        return "Dataset(n=%d, alleles=%s)" % (
+            len(self), self.unique_alleles())
+
+    def __eq__(self, other):
+        """
+        Two datasets are equal if they contain the same number of samples
+        with the same properties and values.
+        """
+        if type(other) is not Dataset:
+            return False
+        elif len(self) != len(other):
+            return False
+        elif len(self.columns) != len(other.columns):
+            return False
+
+        for ci, cj in zip(self.columns, other.columns):
+            if ci != cj:
```

Comment:
Does this work, or throw a "the truth value of an array with more than one entry is ambiguous" error? I'm not sure this is wrong but off the top of my head I would think comparing two series like this would result in that error